put back default rules and use consistent yaml file naming convention

frankbu opened the pull request:
It is a best-practice to set an explicit routing rule for every service early to avoid accidentally deploying a new version and having it immediately receive traffic. I want the bookinfo sample to follow this rule.

All of the yaml files in the bookinfo directory follow a clear and explicit naming convention. I renamed the new one added by @andraxylia to fit in.

@andraxylia Please don't make changes to the bookinfo sample before I've had a chance to review them.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/istio/istio/287)
<!-- Reviewable:end -->

rshriram commented on samples/apps/bookinfo/route-rule-all-v1.yaml:
```diff
@@ -1,10 +1,35 @@
   - tags:
       version: v1
-    weight: 100
 ---
```
May be make the weights explicit to let people know that total weights must add up to 100?

rshriram commented:
@frankbu you need the diffs in the test cases as well.

andraxylia commented:
I get the point of accidentally sending traffic to new versions, but we need to worry more about user understanding the concepts. We structured the bookinfo example unnecessarily complicated for a tutorial. We only demonstrate reviews versioning, but we tagged all services with version one. One may think this is mandatory, when it is not. Bookinfo is not a vanilla app anymore, it is an app where all services have a version tag imposed on them. As a user, I want to try Istio without changing my app specification.
In a real deployment, version label and routing rules can be added as needed. Realistic scenario to be followed when 100 microservices remain unchanged and 1 service gets a new version:

-start with a vanilla app. 
- new versions become available
- add labels to the affected service
   $ kubectl label pods my-pod new-label=v1
- add routing rule to v1
- add the new version with the label v2
- when ready, add routing rule to distribute traffic

istio-testing commented:
Jenkins job istio/presubmit passed

frankbu commented:
I'll add a sentence in the task to explain that setting rules for all services is not necessary but why it is a good idea

andraxylia commented:
If you make version mandatory for services, then the default rule must be added as part of the initial bookinfo deployment. Adding a version should follow the scenario I descried above.

andraxylia commented:
Also, in the absence of weight and the presence of a single version, it is so unclear what the routing rules does. Wight should really be mandatory, the fact that weights add up to 100 is secondary.

andraxylia commented:
Since bookinfo is a tutorial, I propose we eliminate all that is not essential to prove the point, and keep version just for the reviews service, as the accompanying diagram and text describes. We can add a note to recommend similar approach for the other services, but it is not needed to clutter the yaml files and the routing rules with those.
I am also adamant about putting back the weight explicitly, for clarity, and not make it have a default value, for correctness. What if we add a routing rule about something else than the weight, will applying that set the weight to 100?

frankbu commented:
@andraxylia: We've used essentially the same sample in Amalgam8 and I haven't heard of a single piece of feedback that found this confusing. Adding version labels to running pods will likely not be practical in industrial deployments (image 10s or 100s of pods running for a service, including auto-scaling). In fact it may even be impossible in some non-kube deployment environments. Like I said, it is not mandatory, but a best practice. I will add some documentation to make that clear. As far as the weight defaulting to 100 is concerned, I also haven't heard any feedback from anybody other than you that it's confusing. It is well documented. I would argue that if we require it to be set explicitly in rules that have specific match criteria it's even more confusing. For example, a rule that says requests from a specific user should go to v2. Almost always, the intent is 100% of the time the condition is met, but that's not same as 100% of the traffic. Not explicitly having to say weight: 100 doesn't make people think about what is otherwise intuitively obvious.

andraxylia commented:
name: reviews-default
 spec:
   destination: reviews.default.svc.cluster.local
   precedence: 1
   route:
   - tags:
       version: v1

Please tell me what this does, if you come from a world where routing rules are more than percentage of traffic, and do a bunch of other things. Please do not easily ignore new feedback just because you have not heard it before.

frankbu commented:
I'm definitely not ignoring your feedback, but same request in reverse. Please don't assume that your feedback is always representative of the majority of users.

The rule above says that all traffic (100% default weight) for the reviews service should be routed to pods with label "version: v1". Since the precedence is low (1) it is likely intended to be a fall-back rule that applies when no other higher priority rules for the reviews service have been set. Hence, a good name for it is "reviews-default". If for example we want to send traffic from a specific user to different pods (e.g., v2, but really could by any set of tags we want), then we would create another rule with a higher precedence value. How this all works was well documented, but we may have lost some of the information when we cleaned it up and moved from amalgam8 to istio. Here are the docs that we have now:
https://istio.io/docs/concepts/traffic-management/rules-configuration.html
https://istio.io/docs/tasks/request-routing.html

frankbu commented:
Also make sure to read this section: https://istio.io/docs/concepts/traffic-management/rules-configuration.html#rules-have-precedence
